File: product-brain/src/product_brain/adapters/testrail.py
from __future__ import annotations

import base64
from datetime import datetime, timedelta, timezone
from typing import Optional

import httpx

from ..models import RunResult, TestCase
from .test_base import TestAdapter


_AUTOMATION_MAP = {
    0: "manual",
    1: "automated",
    2: "semi",
}


def _ts_to_dt(ts) -> Optional[datetime]:
    if ts in (None, 0, "0"):
        return None
    try:
        return datetime.fromtimestamp(int(ts), tz=timezone.utc)
    except (ValueError, TypeError):
        return None


class TestRailAdapter(TestAdapter):
    """TestRail REST API v2 adapter.

    TestRail uses HTTP Basic Auth (user_email + api_key). Cases are
    linked to tickets via the configurable `refs` field
    (comma-separated).
    """

    def __init__(self, config):
        super().__init__(config)
        cfg = config.testrail
        api_key = config.testrail_api_key() or ""
        token = base64.b64encode(f"{cfg.user_email}:{api_key}".encode()).decode()
        self.base = cfg.base_url.rstrip("/") + "/index.php?/api/v2"
        self._client = httpx.Client(
            headers={
                "Authorization": f"Basic {token}",
                "Content-Type": "application/json",
            },
            timeout=30,
        )
        self.project_id = cfg.project_id
        self.refs_field = cfg.refs_field
        self.window_days = cfg.run_history_window_days

    def _get(self, path: str, **params) -> dict | list:
        sep = "&" if "?" in self.base else "?"
        url = f"{self.base}/{path}"
        r = self._client.get(url, params=params)
        r.raise_for_status()
        return r.json()

    def _to_case(self, raw: dict) -> TestCase:
        case_id = f"TR-C-{raw.get('id', '')}"
        refs = (raw.get(self.refs_field) or "").strip()
        linked = [t.strip() for t in refs.split(",") if t.strip()] if refs else []
        steps_raw = raw.get("custom_steps_separated") or raw.get("custom_steps") or ""
        if isinstance(steps_raw, list):
            steps = [s.get("content", "") for s in steps_raw if isinstance(s, dict)]
        else:
            steps = [s.strip() for s in str(steps_raw).split("\n") if s.strip()]
        return TestCase(
            id=case_id,
            title=raw.get("title", ""),
            preconditions=raw.get("custom_preconds", "") or "",
            steps=steps,
            expected=raw.get("custom_expected", "") or "",
            automation=_AUTOMATION_MAP.get(raw.get("custom_automation_type"), "unknown"),
            type=raw.get("type") or "functional",
            suite=str(raw.get("suite_id", "")),
            linked_tickets=linked,
            url=f"{self.config.testrail.base_url.rstrip('/')}/index.php?/cases/view/{raw.get('id')}",
        )

    def fetch_case(self, case_id: str) -> Optional[TestCase]:
        numeric = case_id.replace("TR-C-", "")
        try:
            data = self._get(f"get_case/{numeric}")
        except httpx.HTTPStatusError:
            return None
        if not isinstance(data, dict):
            return None
        return self._to_case(data)

    def fetch_cases_for_ticket(self, ticket_id: str) -> list[TestCase]:
        if not self.project_id:
            return []
        try:
            data = self._get(f"get_cases/{self.project_id}", refs_filter=ticket_id)
        except httpx.HTTPStatusError:
            return []
        items = data.get("cases", []) if isinstance(data, dict) else (data or [])
        out = []
        for raw in items:
            case = self._to_case(raw)
            if ticket_id in case.linked_tickets:
                out.append(case)
        return out

    def fetch_cases_for_files(self, paths: list[str]) -> list[TestCase]:
        if not paths or not self.project_id:
            return []
        keywords = " ".join({p.split("/")[-1].split(".")[0] for p in paths if p})[:200]
        return self.search_cases(keywords)

    def fetch_run_history(
        self,
        case_id: str,
        since: Optional[datetime] = None,
    ) -> list[RunResult]:
        numeric = case_id.replace("TR-C-", "")
        since = since or (datetime.now(timezone.utc) - timedelta(days=self.window_days))
        cutoff = int(since.timestamp())
        try:
            runs = self._get(f"get_runs/{self.project_id}", created_after=cutoff)
        except httpx.HTTPStatusError:
            return []
        run_items = runs.get("runs", []) if isinstance(runs, dict) else (runs or [])
        out: list[RunResult] = []
        for run in run_items[:50]:
            run_id = run.get("id")
            try:
                results = self._get(f"get_results_for_case/{run_id}/{numeric}")
            except httpx.HTTPStatusError:
                continue
            res_items = results.get("results", []) if isinstance(results, dict) else (results or [])
            for r in res_items:
                out.append(RunResult(
                    case_id=case_id,
                    status=_status_name(r.get("status_id")),
                    run_id=str(run_id),
                    timestamp=_ts_to_dt(r.get("created_on")),
                    comment=(r.get("comment") or "")[:200],
                ))
        return out

    def search_cases(self, keywords: str, limit: int = 50) -> list[TestCase]:
        if not self.project_id:
            return []
        try:
            data = self._get(f"get_cases/{self.project_id}", filter=keywords)
        except httpx.HTTPStatusError:
            return []
        items = data.get("cases", []) if isinstance(data, dict) else (data or [])
        return [self._to_case(c) for c in items[:limit]]


def _status_name(sid) -> str:
    return {
        1: "passed", 2: "blocked", 3: "untested",
        4: "retest", 5: "failed",
    }.get(sid, "unknown")

File: product-brain/src/adapters/testrail.ts
import type { Config } from "../config.js";
import type { RunResult, TestCase } from "../models.js";
import type { TestAdapter } from "./test-base.js";

const AUTOMATION_MAP: Record<number, TestCase["automation"]> = {
  0: "manual",
  1: "automated",
  2: "semi",
};

const STATUS_MAP: Record<number, string> = {
  1: "passed",
  2: "blocked",
  3: "untested",
  4: "retest",
  5: "failed",
};

function tsToDt(ts: unknown): Date | undefined {
  if (ts == null || ts === 0 || ts === "0") return undefined;
  const n = Number(ts);
  if (!Number.isFinite(n)) return undefined;
  return new Date(n * 1000);
}

function asObject(v: unknown): Record<string, unknown> {
  return typeof v === "object" && v !== null ? (v as Record<string, unknown>) : {};
}

export class TestRailAdapter implements TestAdapter {
  private readonly base: string;
  private readonly headers: Record<string, string>;
  private readonly projectId: number;
  private readonly refsField: string;
  private readonly windowDays: number;
  private readonly viewBase: string;

  constructor(private readonly config: Config) {
    const cfg = config.testrail;
    const apiKey = config.testrailApiKey() ?? "";
    const token = Buffer.from(`${cfg.user_email}:${apiKey}`).toString("base64");
    this.viewBase = cfg.base_url.replace(/\/$/, "");
    this.base = `${this.viewBase}/index.php?/api/v2`;
    this.headers = {
      Authorization: `Basic ${token}`,
      "Content-Type": "application/json",
    };
    this.projectId = cfg.project_id;
    this.refsField = cfg.refs_field;
    this.windowDays = cfg.run_history_window_days;
  }

  private async req(path: string, params?: Record<string, string | number>): Promise<unknown> {
    const url = new URL(`${this.base}/${path}`);
    if (params) {
      for (const [k, v] of Object.entries(params)) url.searchParams.set(k, String(v));
    }
    const r = await fetch(url, {
      headers: this.headers,
      signal: AbortSignal.timeout(30_000),
    });
    if (!r.ok) throw new Error(`TestRail GET ${path} → ${r.status}`);
    return r.json();
  }

  private toCase(raw: Record<string, unknown>): TestCase {
    const id = `TR-C-${raw["id"] ?? ""}`;
    const refs = String(raw[this.refsField] ?? "").trim();
    const linkedTickets = refs ? refs.split(",").map((t) => t.trim()).filter(Boolean) : [];
    const stepsRaw = raw["custom_steps_separated"] ?? raw["custom_steps"] ?? "";
    let steps: string[];
    if (Array.isArray(stepsRaw)) {
      steps = stepsRaw
        .filter((s): s is Record<string, unknown> => typeof s === "object" && s !== null)
        .map((s) => String(s["content"] ?? ""));
    } else {
      steps = String(stepsRaw).split("\n").map((s) => s.trim()).filter(Boolean);
    }
    const automationRaw = raw["custom_automation_type"];
    const automation =
      typeof automationRaw === "number" ? (AUTOMATION_MAP[automationRaw] ?? "unknown") : "unknown";
    return {
      id,
      title: String(raw["title"] ?? ""),
      preconditions: String(raw["custom_preconds"] ?? ""),
      steps,
      expected: String(raw["custom_expected"] ?? ""),
      automation,
      type: String(raw["type"] ?? "functional"),
      suite: String(raw["suite_id"] ?? ""),
      linkedTickets,
      recentFailures: 0,
      url: `${this.viewBase}/index.php?/cases/view/${raw["id"] ?? ""}`,
    };
  }

  async fetchCase(caseId: string): Promise<TestCase | null> {
    const numeric = caseId.replace("TR-C-", "");
    try {
      const data = await this.req(`get_case/${numeric}`);
      if (typeof data !== "object" || data === null) return null;
      return this.toCase(data as Record<string, unknown>);
    } catch {
      return null;
    }
  }

  async fetchCasesForTicket(ticketId: string): Promise<TestCase[]> {
    if (!this.projectId) return [];
    let raw: unknown;
    try {
      raw = await this.req(`get_cases/${this.projectId}`, { refs_filter: ticketId });
    } catch {
      return [];
    }
    const items = Array.isArray(raw)
      ? (raw as Array<Record<string, unknown>>)
      : ((asObject(raw)["cases"] as Array<Record<string, unknown>>) ?? []);
    return items
      .map((r) => this.toCase(r))
      .filter((c) => c.linkedTickets.includes(ticketId));
  }

  async fetchCasesForFiles(paths: string[]): Promise<TestCase[]> {
    if (!paths.length || !this.projectId) return [];
    const keywords = [...new Set(paths.map((p) => p.split("/").pop()?.split(".")[0]).filter(Boolean))]
      .join(" ")
      .slice(0, 200);
    return this.searchCases(keywords);
  }

  async fetchRunHistory(caseId: string, since?: Date): Promise<RunResult[]> {
    const numeric = caseId.replace("TR-C-", "");
    const sinceDt = since ?? new Date(Date.now() - this.windowDays * 86_400_000);
    const cutoff = Math.floor(sinceDt.getTime() / 1000);
    let runs: unknown;
    try {
      runs = await this.req(`get_runs/${this.projectId}`, { created_after: cutoff });
    } catch {
      return [];
    }
    const runItems = Array.isArray(runs)
      ? (runs as Array<Record<string, unknown>>)
      : ((asObject(runs)["runs"] as Array<Record<string, unknown>>) ?? []);
    const out: RunResult[] = [];
    for (const run of runItems.slice(0, 50)) {
      const runId = run["id"];
      let results: unknown;
      try {
        results = await this.req(`get_results_for_case/${runId}/${numeric}`);
      } catch {
        continue;
      }
      const resItems = Array.isArray(results)
        ? (results as Array<Record<string, unknown>>)
        : ((asObject(results)["results"] as Array<Record<string, unknown>>) ?? []);
      for (const r of resItems) {
        const sid = r["status_id"];
        out.push({
          caseId,
          status: typeof sid === "number" ? (STATUS_MAP[sid] ?? "unknown") : "unknown",
          runId: String(runId ?? ""),
          timestamp: tsToDt(r["created_on"]),
          comment: String(r["comment"] ?? "").slice(0, 200),
        });
      }
    }
    return out;
  }

  async searchCases(keywords: string, limit = 50): Promise<TestCase[]> {
    if (!this.projectId) return [];
    let raw: unknown;
    try {
      raw = await this.req(`get_cases/${this.projectId}`, { filter: keywords });
    } catch {
      return [];
    }
    const items = Array.isArray(raw)
      ? (raw as Array<Record<string, unknown>>)
      : ((asObject(raw)["cases"] as Array<Record<string, unknown>>) ?? []);
    return items.slice(0, limit).map((r) => this.toCase(r));
  }
}
